File: 985-bag-of-tokens/bag-of-tokens.scala
object Solution {
    def bagOfTokensScore(tokens: Array[Int], pow: Int): Int = {
        var score, lost, l = 0
        var power = pow
        var r = tokens.length -1

        tokens.sortInPlace

        while (l<=r && (score != 0 || power >= tokens(l)))
            if (score == 0 || power >= tokens(l)) {
                power -= tokens(l)
                score += 1
                lost = 0
                l+=1
            } else {
                power += tokens(r)
                score -= 1
                lost += 1
                r -= 1
            }
        score + lost    
    }
}

File: 985-bag-of-tokens/bag-of-tokens.py
class Solution:
    def bagOfTokensScore(self, tokens: List[int], power: int) -> int:
        tokens.sort()
        score = 0
        a, b = 0, len(tokens)-1

        while a <= b:
            if power >= tokens[a]:
                power -= tokens[a]
                a+=1
                score+=1
            elif b >= a+1 and score >0:
                power += tokens[b]
                score -= 1
                b -=1
            else:
                break
        return score